File: js/packages/types/src/questNode.ts
import { PointOfInterest } from './pointOfInterest';

export type QuestNodeSubmissionType =
  | 'text'
  | 'photo'
  | 'video'
  | (string & {});

export type QuestNodeObjectiveType =
  | 'challenge'
  | 'scenario'
  | 'monster_encounter'
  | 'monster'
  | (string & {});

export interface QuestNodeObjective {
  id: string;
  type: QuestNodeObjectiveType;
  prompt: string;
  description?: string;
  imageUrl?: string;
  thumbnailUrl?: string;
  reward?: number;
  inventoryItemId?: number | null;
  submissionType?: QuestNodeSubmissionType;
  difficulty?: number;
  statTags?: string[];
  proficiency?: string | null;
}

export interface QuestNodeChallengeDetails {
  id: string;
  pointOfInterestId?: string | null;
  pointOfInterest?: PointOfInterest | null;
  latitude: number;
  longitude: number;
  question: string;
  description?: string;
  submissionType?: QuestNodeSubmissionType;
  difficulty?: number;
  statTags?: string[];
  proficiency?: string | null;
  reward?: number;
  rewardExperience?: number;
}

export interface QuestNodeScenarioOptionDetails {
  id: string;
}

export interface QuestNodeScenarioDetails {
  id: string;
  zoneId: string;
  pointOfInterestId?: string | null;
  pointOfInterest?: PointOfInterest | null;
  latitude: number;
  longitude: number;
  prompt: string;
  difficulty?: number;
  openEnded?: boolean;
  options?: QuestNodeScenarioOptionDetails[];
}

export interface QuestNodeMonsterMemberDetails {
  slot: number;
  monster: { id: string; name: string };
}

export interface QuestNodeMonsterEncounterDetails {
  id: string;
  zoneId: string;
  pointOfInterestId?: string | null;
  pointOfInterest?: PointOfInterest | null;
  latitude: number;
  longitude: number;
  name: string;
  description?: string;
  encounterType?: string;
  scaleWithUserLevel?: boolean;
  monsterCount?: number;
  members?: QuestNodeMonsterMemberDetails[];
}

export interface QuestNodeMonsterDetails {
  id: string;
  zoneId?: string;
  latitude: number;
  longitude: number;
  name: string;
  description?: string;
  level?: number;
}

export interface QuestNode {
  id: string;
  questId: string;
  orderIndex: number;
  submissionType?: QuestNodeSubmissionType;
  objectiveText?: string;
  objective?: QuestNodeObjective | null;
  pointOfInterestId?: string | null;
  scenarioId?: string | null;
  monsterId?: string | null;
  monsterEncounterId?: string | null;
  challengeId?: string | null;
  polygon?: string | null;
  polygonPoints?: [number, number][];
  scenario?: QuestNodeScenarioDetails | null;
  monsterEncounter?: QuestNodeMonsterEncounterDetails | null;
  monster?: QuestNodeMonsterDetails | null;
  challenge?: QuestNodeChallengeDetails | null;
}

File: dart/unclaimed_streets/lib/models/quest_node.dart
import 'point_of_interest.dart';
import 'exposition.dart';
import 'quest_node_objective.dart';

class QuestNodePolygonPoint {
  final double latitude;
  final double longitude;

  const QuestNodePolygonPoint({
    required this.latitude,
    required this.longitude,
  });

  factory QuestNodePolygonPoint.fromJson(Map<String, dynamic> json) {
    return QuestNodePolygonPoint(
      latitude: (json['latitude'] as num?)?.toDouble() ?? 0.0,
      longitude: (json['longitude'] as num?)?.toDouble() ?? 0.0,
    );
  }
}

class QuestNode {
  static const submissionTypeText = 'text';
  static const submissionTypePhoto = 'photo';
  static const submissionTypeVideo = 'video';

  final String id;
  final int orderIndex;
  final PointOfInterest? pointOfInterest;
  final String objectiveText;
  final String? scenarioId;
  final String? expositionId;
  final String? monsterId;
  final String? monsterEncounterId;
  final String? challengeId;
  final List<QuestNodePolygonPoint> polygon;
  final QuestNodeObjective? objective;
  final Exposition? exposition;
  final String submissionType;

  const QuestNode({
    required this.id,
    required this.orderIndex,
    this.submissionType = submissionTypePhoto,
    this.pointOfInterest,
    this.objectiveText = '',
    this.scenarioId,
    this.expositionId,
    this.monsterId,
    this.monsterEncounterId,
    this.challengeId,
    this.polygon = const [],
    this.objective,
    this.exposition,
  });

  factory QuestNode.fromJson(Map<String, dynamic> json) {
    final rawSubmissionType = (json['submissionType'] as String?)
        ?.trim()
        .toLowerCase();
    final submissionType =
        (rawSubmissionType == submissionTypeText ||
            rawSubmissionType == submissionTypePhoto ||
            rawSubmissionType == submissionTypeVideo)
        ? rawSubmissionType!
        : submissionTypePhoto;
    return QuestNode(
      id: json['id'] as String? ?? '',
      orderIndex: (json['orderIndex'] as num?)?.toInt() ?? 0,
      submissionType: submissionType,
      pointOfInterest: json['pointOfInterest'] is Map<String, dynamic>
          ? PointOfInterest.fromJson(
              json['pointOfInterest'] as Map<String, dynamic>,
            )
          : null,
      objectiveText: json['objectiveText']?.toString() ?? '',
      scenarioId: json['scenarioId']?.toString(),
      expositionId: json['expositionId']?.toString(),
      monsterId: json['monsterId']?.toString(),
      monsterEncounterId: json['monsterEncounterId']?.toString(),
      challengeId: json['challengeId']?.toString(),
      polygon:
          (json['polygon'] as List<dynamic>?)
              ?.map(
                (e) =>
                    QuestNodePolygonPoint.fromJson(e as Map<String, dynamic>),
              )
              .toList() ??
          const [],
      objective: json['objective'] is Map<String, dynamic>
          ? QuestNodeObjective.fromJson(
              json['objective'] as Map<String, dynamic>,
            )
          : null,
      exposition: json['exposition'] is Map<String, dynamic>
          ? Exposition.fromJson(json['exposition'] as Map<String, dynamic>)
          : json['exposition'] is Map
          ? Exposition.fromJson(
              Map<String, dynamic>.from(json['exposition'] as Map),
            )
          : null,
    );
  }
}
